### DISCUSSION TITLE
OpenVDB AX Merge

### DISCUSSION THREAD

**Discussion by chrisflesher** [AcademySoftwareFoundation/openvdb#1580] [Q&A] [answered]
I am trying to merge two level sets that have non-overlapping active areas. There is some [documentation here](https://www.openvdb.org/documentation/doxygen/axexamples.html) (in the "Linear Blending" section) that seems to suggest this is possible.

Here are some things I tried... but so far have been unsuccessful. One of the grids has a rotational transform applied so I can't easily copy active indices from one grid to another (e.g. using python). Any suggestions would be appreciated!
```
$ vdb_print input.vdb 
surface_a    float (-41,40,-1)->(59,141,51)      101x102x53   285KVox 1.41MB
surface_b    float (-51,71,-1)->(50,171,50)      102x101x52   256KVox 1.39MB

$ vdb_ax input.vdb output.vdb -s "float@surface_a = min(float@surface_a, float@surface_b);"
$ vdb_print output.vdb
surface_a    float (-41,40,-1)->(59,141,51)      101x102x53   285KVox 1.41MB
surface_b    float (-51,71,-1)->(50,171,50)      102x101x52   256KVox 1.39MB

$ vdb_ax input.vdb output.vdb -s "float@surface_c = min(float@surface_a, float@surface_b);"
$ vdb_print output.vdb
surface_a    float (-41,40,-1)->(59,141,51)      101x102x53   285KVox 1.41MB
surface_b    float (-51,71,-1)->(50,171,50)      102x101x52   256KVox 1.39MB
surface_c    float (2147483647,2147483647,2147483647)->(-2147483648,-2147483648,-2147483648) 0x0x0      0Vox 1.26KB

$ vdb_print --version
OpenVDB library version: 8.1.0abi8
OpenVDB file format version: 224
```

**Comment by Idclip** [comment, marked as answer]
Hey so the short answer is that you can't do this solely with theb `vdb_ax` command line binary, but I think I will add an option to allow it because the ideal solution has been proposed for a while but is still not completed. The option will most likely manifest as a `--resample-to-match` or `--combine` flag.

--------------------

Some more details - The issue is that when you're dealing with volumes of different transforms, an expression like `min(float@surface_a, float@surface_b)` becomes a bit overloaded - attribute accesses should ideally be delegating automatically to more accurate (i.e. trilinear) samplings to accomodate the differences in resolution with respect to the target volume, but at the moment these accesses are always point samples at ijk coordinates.

A somewhat naive but valid approach is to resample volumes to the target volumes transform. In C++ this would look something like:

```
FloatGrid a, b ....;

FloatGrid resampled_b;
resampled_b.setName("b");

// Resample b
resampled_b.setTransform(a.transformPtr());
openvdb::tools::resampleToMatch<openvdb::tools::BoxSampler>(b, *resampled_b);

// Now activate it where it needs to be
b.topologyUnion(a);

openvdb::ax::run("float@surface_a = min(float@surface_a, float@surface_b);",  grids);
```

Obviously if your transforms match then this just becomes:

```
FloatGrid a, b ....;
b.topologyUnion(a);
openvdb::ax::run("float@surface_a = min(float@surface_a, float@surface_b);",  grids);
```

This is basically what I think I'll add to start with. The reason this behaviour hasn't historically been part of AX (i.e. you need to call resample/topologyUnion methods) is that, should resampling be necessary, it can be very expensive. If we instead resampled on the fly, we would be able to get away without resampling the whole volume and only resample when the expression is evaluated. But it's only mis-matching transforms that are the issue - a topologyUnion with matching transforms is extremely fast and is most likely the more common requirement.

**Reply by chrisflesher** [reply]
Thanks for the answer! I like the `vdb_ax --resample-to-match` idea. If its easy to implement it might be nice to have different interpolation options (e.g. nearest, trilinear, and tricubic).
